File: packages/core/src/nn/transformer.ts
import { PipelineManager } from "../gpu/pipeline.js";
import { BufferPool, createUniformBuffer } from "../gpu/buffer-pool.js";
import { Attention } from "./attention.js";
import { FFN } from "./ffn.js";
import { type BindGroupCache, createBGCache, clearBGCache, cachedBG } from "./bg-cache.js";
import type { ModelConfig, KVCache } from "../types.js";

import rmsnormWGSL from "../shaders/rmsnorm.wgsl";
import elementwiseWGSL from "../shaders/elementwise.wgsl";

/**
 * Single transformer block:
 *   residual = x
 *   x = attention(input_layernorm(x)) + residual
 *   residual = x
 *   x = ffn(post_attention_layernorm(x)) + residual
 */
export class TransformerBlock {
  private device: GPUDevice;
  private pipelines: PipelineManager;
  private pool: BufferPool;
  private config: ModelConfig;

  private inputLayerNorm: GPUBuffer; // [hidden] f32
  private postAttnLayerNorm: GPUBuffer; // [hidden] f32
  private attention: Attention;
  private ffn: FFN;

  // Pre-created uniform buffers for N=1 decode (static params, shared within block)
  private decodeNormUniform?: GPUBuffer;
  private decodeAddUniform?: GPUBuffer;

  // Pre-created uniform buffers for N>1 prefill (dynamic — updated via writeBuffer)
  private prefillNormUniform?: GPUBuffer;
  private prefillAddUniform?: GPUBuffer;

  // Bind group cache for N=1 decode
  private bgCache: BindGroupCache = createBGCache();

  constructor(
    device: GPUDevice,
    pipelines: PipelineManager,
    pool: BufferPool,
    config: ModelConfig,
    inputLayerNorm: GPUBuffer,
    postAttnLayerNorm: GPUBuffer,
    attention: Attention,
    ffn: FFN
  ) {
    this.device = device;
    this.pipelines = pipelines;
    this.pool = pool;
    this.config = config;
    this.inputLayerNorm = inputLayerNorm;
    this.postAttnLayerNorm = postAttnLayerNorm;
    this.attention = attention;
    this.ffn = ffn;
  }

  /** Pre-create uniform buffers for N=1 decode path (all static). */
  initDecodeUniforms(maxSeqLen: number): void {
    {
      const data = new ArrayBuffer(12);
      const v = new DataView(data);
      v.setUint32(0, 1, true);
      v.setUint32(4, this.config.hiddenSize, true);
      v.setFloat32(8, this.config.rmsNormEps, true);
      this.decodeNormUniform = createUniformBuffer(this.device, data);
    }
    {
      const data = new ArrayBuffer(8);
      const v = new DataView(data);
      v.setUint32(0, this.config.hiddenSize, true);
      v.setUint32(4, 0, true); // add
      this.decodeAddUniform = createUniformBuffer(this.device, data);
    }

    // Prefill uniforms (reused via writeBuffer for N>1)
    const mkBuf = (size: number) =>
      this.device.createBuffer({
        size,
        usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
      });
    this.prefillNormUniform = mkBuf(12);
    this.prefillAddUniform = mkBuf(8);

    this.attention.initDecodeUniforms(maxSeqLen);
    this.ffn.initDecodeUniforms();
  }

  /**
   * Forward pass through one transformer block.
   * @param input [N, hidden] f32
   * @param N sequence length
   * @param kvCache KV cache for this layer
   * @param encoder GPU command encoder
   * @returns output [N, hidden] f32
   */
  forward(
    input: GPUBuffer,
    N: number,
    kvCache: KVCache,
    encoder: GPUCommandEncoder
  ): GPUBuffer {
    const hidden = this.config.hiddenSize;

    // Self-attention with residual
    const normedAttn = this.dispatchRMSNorm(
      encoder,
      input,
      this.inputLayerNorm,
      N,
      "attnNorm"
    );
    const attnOut = this.attention.forward(normedAttn, N, kvCache, encoder);
    this.pool.release(normedAttn);

    const residual1 = this.dispatchAdd(
      encoder,
      input,
      attnOut,
      N * hidden,
      N,
      "attnAdd"
    );
    this.pool.release(attnOut);

    // FFN with residual
    const normedFFN = this.dispatchRMSNorm(
      encoder,
      residual1,
      this.postAttnLayerNorm,
      N,
      "ffnNorm"
    );
    const ffnOut = this.ffn.forward(normedFFN, N, encoder);
    this.pool.release(normedFFN);

    const output = this.dispatchAdd(
      encoder,
      residual1,
      ffnOut,
      N * hidden,
      N,
      "ffnAdd"
    );
    this.pool.release(residual1);
    this.pool.release(ffnOut);

    return output;
  }

  private dispatchRMSNorm(
    encoder: GPUCommandEncoder,
    input: GPUBuffer,
    weight: GPUBuffer,
    N: number,
    bgId?: string
  ): GPUBuffer {
    const { pipeline, bindGroupLayout } = this.pipelines.getOrCreate(
      "rmsnorm",
      rmsnormWGSL
    );

    const hidden = this.config.hiddenSize;
    const output = this.pool.acquire(
      N * hidden * 4,
      GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC
    );

    let paramsBuffer: GPUBuffer;
    if (N === 1 && this.decodeNormUniform) {
      paramsBuffer = this.decodeNormUniform;
    } else if (this.prefillNormUniform) {
      const paramsData = new ArrayBuffer(12);
      const v = new DataView(paramsData);
      v.setUint32(0, N, true);
      v.setUint32(4, hidden, true);
      v.setFloat32(8, this.config.rmsNormEps, true);
      this.device.queue.writeBuffer(this.prefillNormUniform, 0, new Uint8Array(paramsData));
      paramsBuffer = this.prefillNormUniform;
    } else {
      const paramsData = new ArrayBuffer(12);
      const v = new DataView(paramsData);
      v.setUint32(0, N, true);
      v.setUint32(4, hidden, true);
      v.setFloat32(8, this.config.rmsNormEps, true);
      paramsBuffer = createUniformBuffer(this.device, paramsData);
    }

    const entries: GPUBindGroupEntry[] = [
      { binding: 0, resource: { buffer: input } },
      { binding: 1, resource: { buffer: weight } },
      { binding: 2, resource: { buffer: output } },
      { binding: 3, resource: { buffer: paramsBuffer } },
    ];
    const bindGroup = N === 1 && bgId
      ? cachedBG(this.bgCache, this.device, bgId, bindGroupLayout, entries)
      : this.device.createBindGroup({ layout: bindGroupLayout, entries });

    const pass = encoder.beginComputePass();
    pass.setPipeline(pipeline);
    pass.setBindGroup(0, bindGroup);
    pass.dispatchWorkgroups(N);
    pass.end();

    return output;
  }

  private dispatchAdd(
    encoder: GPUCommandEncoder,
    a: GPUBuffer,
    b: GPUBuffer,
    numElements: number,
    N: number,
    bgId?: string
  ): GPUBuffer {
    const { pipeline, bindGroupLayout } = this.pipelines.getOrCreate(
      "elementwise_0",
      elementwiseWGSL
    );

    const output = this.pool.acquire(
      numElements * 4,
      GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC
    );

    let paramsBuffer: GPUBuffer;
    if (N === 1 && this.decodeAddUniform) {
      paramsBuffer = this.decodeAddUniform;
    } else if (this.prefillAddUniform) {
      const paramsData = new ArrayBuffer(8);
      const v = new DataView(paramsData);
      v.setUint32(0, numElements, true);
      v.setUint32(4, 0, true); // add
      this.device.queue.writeBuffer(this.prefillAddUniform, 0, new Uint8Array(paramsData));
      paramsBuffer = this.prefillAddUniform;
    } else {
      const paramsData = new ArrayBuffer(8);
      const v = new DataView(paramsData);
      v.setUint32(0, numElements, true);
      v.setUint32(4, 0, true); // add
      paramsBuffer = createUniformBuffer(this.device, paramsData);
    }

    const entries: GPUBindGroupEntry[] = [
      { binding: 0, resource: { buffer: a } },
      { binding: 1, resource: { buffer: b } },
      { binding: 2, resource: { buffer: output } },
      { binding: 3, resource: { buffer: paramsBuffer } },
    ];
    const bindGroup = N === 1 && bgId
      ? cachedBG(this.bgCache, this.device, bgId, bindGroupLayout, entries)
      : this.device.createBindGroup({ layout: bindGroupLayout, entries });

    const pass = encoder.beginComputePass();
    pass.setPipeline(pipeline);
    pass.setBindGroup(0, bindGroup);
    pass.dispatchWorkgroups(Math.ceil(numElements / 256));
    pass.end();

    return output;
  }

  /** Clear the bind group cache (call on KV cache reset). */
  clearBGCache(): void {
    clearBGCache(this.bgCache);
    this.attention.clearBGCache();
    this.ffn.clearBGCache();
  }

  /** Destroy pre-allocated attention buffers. */
  destroyPreAllocated(): void {
    this.attention.destroyPreAllocated();
  }

}

File: packages/rust/crates/oxbitnet/src/nn/transformer.rs
use std::sync::Arc;

use wgpu::BufferUsages;

use crate::gpu::buffer_pool::{BufferPool, GpuBuf};
use crate::gpu::pipeline::PipelineManager;
use crate::model::config::ModelConfig;
use crate::nn::attention::{Attention, KvCache};
use crate::nn::bg_cache::BgCache;
use crate::nn::bitlinear::{buf_entry, create_uniform_raw, create_uniform_u32_u32_f32};
use crate::nn::ffn::FFN;

const RMSNORM_WGSL: &str = include_str!("../shaders/rmsnorm.wgsl");
const ELEMENTWISE_WGSL: &str = include_str!("../shaders/elementwise.wgsl");

/// Single transformer block:
///   residual = x
///   x = attention(input_layernorm(x)) + residual
///   residual = x
///   x = ffn(post_attention_layernorm(x)) + residual
pub struct TransformerBlock {
    device: Arc<wgpu::Device>,
    config: ModelConfig,
    input_layer_norm: GpuBuf,
    post_attn_layer_norm: GpuBuf,
    pub(crate) attention: Attention,
    pub(crate) ffn: FFN,
    bg_cache: BgCache,
}

impl TransformerBlock {
    pub fn new(
        device: Arc<wgpu::Device>,
        config: ModelConfig,
        input_layer_norm: GpuBuf,
        post_attn_layer_norm: GpuBuf,
        attention: Attention,
        ffn: FFN,
    ) -> Self {
        Self {
            device,
            config,
            input_layer_norm,
            post_attn_layer_norm,
            attention,
            ffn,
            bg_cache: BgCache::new(),
        }
    }

    pub fn forward(
        &mut self,
        input: &GpuBuf,
        n: usize,
        kv_cache: &mut KvCache,
        encoder: &mut wgpu::CommandEncoder,
        pipelines: &mut PipelineManager,
        pool: &BufferPool,
    ) -> GpuBuf {
        let hidden = self.config.hidden_size;

        // Self-attention with residual
        let normed_attn = self.dispatch_rmsnorm(encoder, input, &self.input_layer_norm.clone(), n, pipelines, pool);
        let attn_out = self.attention.forward(&normed_attn, n, kv_cache, encoder, pipelines, pool);

        let residual1 = self.dispatch_add(encoder, input, &attn_out, n * hidden, pipelines, pool);

        // FFN with residual
        let normed_ffn = self.dispatch_rmsnorm(encoder, &residual1, &self.post_attn_layer_norm.clone(), n, pipelines, pool);
        let ffn_out = self.ffn.forward(&normed_ffn, n, encoder, pipelines, pool);

        let output = self.dispatch_add(encoder, &residual1, &ffn_out, n * hidden, pipelines, pool);

        output
    }

    fn dispatch_rmsnorm(
        &self,
        encoder: &mut wgpu::CommandEncoder,
        input: &wgpu::Buffer,
        weight: &wgpu::Buffer,
        n: usize,
        pipelines: &mut PipelineManager,
        pool: &BufferPool,
    ) -> GpuBuf {
        let entry = pipelines.get_or_create_default("rmsnorm", RMSNORM_WGSL);
        let hidden = self.config.hidden_size;

        let output = pool.acquire(
            (n * hidden * 4) as u64,
            BufferUsages::STORAGE | BufferUsages::COPY_SRC,
        );

        let params = create_uniform_u32_u32_f32(
            &self.device,
            n as u32,
            hidden as u32,
            self.config.rms_norm_eps,
        );

        let bg = self.device.create_bind_group(&wgpu::BindGroupDescriptor {
            label: Some("transformer_rmsnorm"),
            layout: &entry.bind_group_layout,
            entries: &[
                buf_entry(0, input),
                buf_entry(1, weight),
                buf_entry(2, &output),
                buf_entry(3, &params),
            ],
        });

        let mut pass = encoder.begin_compute_pass(&Default::default());
        pass.set_pipeline(&entry.pipeline);
        pass.set_bind_group(0, Some(&bg), &[]);
        pass.dispatch_workgroups(n as u32, 1, 1);

        output
    }

    fn dispatch_add(
        &self,
        encoder: &mut wgpu::CommandEncoder,
        a: &wgpu::Buffer,
        b: &wgpu::Buffer,
        num_elements: usize,
        pipelines: &mut PipelineManager,
        pool: &BufferPool,
    ) -> GpuBuf {
        let entry = pipelines.get_or_create_default("elementwise_0", ELEMENTWISE_WGSL);

        let output = pool.acquire(
            (num_elements * 4) as u64,
            BufferUsages::STORAGE | BufferUsages::COPY_SRC,
        );

        let params_data = [
            (num_elements as u32).to_le_bytes(),
            0u32.to_le_bytes(), // add
        ]
        .concat();
        let params = create_uniform_raw(&self.device, &params_data);

        let bg = self.device.create_bind_group(&wgpu::BindGroupDescriptor {
            label: Some("transformer_add"),
            layout: &entry.bind_group_layout,
            entries: &[
                buf_entry(0, a),
                buf_entry(1, b),
                buf_entry(2, &output),
                buf_entry(3, &params),
            ],
        });

        let mut pass = encoder.begin_compute_pass(&Default::default());
        pass.set_pipeline(&entry.pipeline);
        pass.set_bind_group(0, Some(&bg), &[]);
        pass.dispatch_workgroups(((num_elements + 255) / 256) as u32, 1, 1);

        output
    }

    pub fn clear_bg_cache(&mut self) {
        self.bg_cache.clear();
        self.attention.clear_bg_cache();
        self.ffn.clear_bg_cache();
    }
}
